src/Printer/CollectorMetadataPrinter.php

```diff
@@ -80,29 +82,51 @@ public function printArgTypesAsString(MethodCall $methodCall, ExtendedMethodRefl
         return implode('|', $stringArgTypes);
     }
 
-    public function printParamTypesToString(ClassMethod $classMethod, ?string $className): string
+    public function printParamTypesToString(ClassMethod $classMethod, ?string $className, ClassReflection $classReflection, Scope $scope): string
     {
+        $parametersReflection = [];
+        if ($classReflection->hasMethod($classMethod->name->name)) {
+            $methodReflection = $classReflection->getMethod($classMethod->name->name, $scope);
+            $variants = $methodReflection->getVariants();
+            if (count($variants) === 1) {
+                $parametersReflection = $variants[0]->getParameters();
+            }
+        }
+
         $printedParamTypes = [];
-        foreach ($classMethod->params as $param) {
+        foreach ($classMethod->params as $i => $param) {
             if ($param->type === null) {
                 $printedParamTypes[] = '';
                 continue;
             }
 
-            $paramType = $this->transformSelfToClassName($param->type, $className);
-            if ($paramType instanceof NullableType) {
-                // unite to phpstan type
-                $paramType = new UnionType([$paramType->type, new Identifier('null')]);
+            $phpdocType = null;
+            if (array_key_exists($i, $parametersReflection)) {
+                $paramphpdocType = $parametersReflection[$i]->getPhpDocType();
+                if (!$paramphpdocType instanceof MixedType) {
+                    $phpdocType = $paramphpdocType;
+                }
```

Comment:
might make sense to scope this fix on Generic types for now, as I am not sure how things like more precise string types e.g. `non-empty-string` fit together with the mapping the lib is doing